File: backend-quarkus/src/main/java/app/nanotes/backend/web/CurrentSession.java
package app.nanotes.backend.web;

import app.nanotes.backend.apperr.UnauthorizedException;
import jakarta.enterprise.context.RequestScoped;

/**
 * Holds the session resolved from the request's cookie (if any), populated
 * once per request by {@link SessionResolutionFilter}. Mirrors backend-go's
 * context values set by sessionContextMiddleware.
 */
@RequestScoped
public class CurrentSession {

    private String sessionId;
    private String userId;
    private String csrfToken;

    void set(String sessionId, String userId, String csrfToken) {
        this.sessionId = sessionId;
        this.userId = userId;
        this.csrfToken = csrfToken;
    }

    /** The raw session cookie value, or null for an anonymous caller. */
    public String sessionId() {
        return sessionId;
    }

    /** The caller's user ID, or null for an anonymous caller. */
    public String userId() {
        return userId;
    }

    public String csrfToken() {
        return csrfToken;
    }

    public boolean isAuthenticated() {
        return userId != null;
    }

    /** Returns the caller's user ID, or throws {@link UnauthorizedException} for an anonymous caller. */
    public String requireUserId() {
        if (userId == null) {
            throw new UnauthorizedException();
        }
        return userId;
    }
}

File: backend-quarkus/src/main/kotlin/app/nanotes/backend/web/CurrentSession.kt
package app.nanotes.backend.web

import app.nanotes.backend.apperr.UnauthorizedException
import jakarta.enterprise.context.RequestScoped

/**
 * Holds the session resolved from the request's cookie (if any), populated
 * once per request by [SessionResolutionFilter]. Mirrors backend-go's
 * context values set by sessionContextMiddleware.
 *
 * Exposed as accessor functions rather than `var` properties: this class is
 * `@RequestScoped`, so Quarkus's CDI container generates an open subclass
 * proxy for it (see the `allOpen` config in build.gradle.kts) — an open
 * `var` with a `private set` is invalid Kotlin, since a proxy would need to
 * override the setter too.
 */
@RequestScoped
class CurrentSession {
    private var sessionIdValue: String? = null
    private var userIdValue: String? = null
    private var csrfTokenValue: String? = null

    fun set(sessionId: String, userId: String, csrfToken: String) {
        sessionIdValue = sessionId
        userIdValue = userId
        csrfTokenValue = csrfToken
    }

    /** The raw session cookie value, or null for an anonymous caller. */
    fun sessionId(): String? = sessionIdValue

    /** The caller's user ID, or null for an anonymous caller. */
    fun userId(): String? = userIdValue

    fun csrfToken(): String? = csrfTokenValue

    val isAuthenticated: Boolean
        get() = userIdValue != null

    /** Returns the caller's user ID, or throws [UnauthorizedException] for an anonymous caller. */
    fun requireUserId(): String = userIdValue ?: throw UnauthorizedException()
}
